src/xdsl/rewriter.py

```diff
@@ -40,10 +40,11 @@ class RewritePattern(ABC):
     @abstractmethod
     def match_and_rewrite(
             self, op: Operation,
-            new_operands: List[SSAValue]) -> Optional[RewriteAction]:
+            new_operands: Optional[List[SSAValue]]) -> Optional[RewriteAction]:
```

Comment:
Are you sure that `Optional[List[SSAValue]]` is a good idea? How is this different from a potential empty list, i.e., how is the difference in semantics?
Furthermore, the walker calls `match_and_rewrite` with the result of `get_new_operands`, which can never return `None`, but it can return a list containing `None`, thus this should actually be `List[Optional[SSAValue]]`.

Reply:
Yeah, I messed up while modifying the types. It is indeed supposed to be `List[Optional[SSAValue]]`